src/charm.py

```diff
@@ -38,45 +44,61 @@ def __init__(self, *args):
         if not self.unit.is_leader():
             return
 
-        # do not initialise replica set until all peers have joined
-        if not self._check_unit_count:
-            logger.debug("waiting to initialise replica set until all planned units have joined")
+        # app-changed-events can trigger the relation changed hook resulting in no JUJU_REMOTE_UNIT
+        if event.unit is None:
             return
 
-        # do not initialise replica set until all peers have started mongod
-        for peer in self._peers.units:
-            mongo_peer = self._single_mongo_replica(
-                str(self._peers.data[peer].get("private-address"))
+        #  only add the calling unit to the replica set if it has mongod running
+        calling_unit = event.unit
+        calling_unit_ip = str(self._peers.data[calling_unit].get("private-address"))
+        mongo_peer = self._single_mongo_replica(calling_unit_ip)
+        if not mongo_peer.is_ready(standalone=True):
+            logger.debug(
+                "unit is not ready, cannot initialise replica set unit: %s is ready, deferring on relation-joined",
+                calling_unit.name,
             )
-            if not mongo_peer.is_ready(standalone=True):
-                logger.debug(
-                    "unit: %s is not ready, cannot initialise replica set until all units are ready, deferring on relation-joined",
-                    peer,
-                )
-                event.defer()
-                return
+            event.defer()
+            return
 
-        # in future patch we will reconfigure the replicaset instead of re-initialising
-        if not self._mongo.is_replica_set():
-            try:
-                self._initialise_replica_set()
-            except (ConnectionFailure, ConfigurationError):
-                self.unit.status = BlockedStatus("failed to initialise replica set")
-                return
+        logger.debug("unit: %s is ready to join the replica set", calling_unit.name)
+        self._reconfigure(event)
 
-        # verify replica set is ready
-        if not self._mongo.is_ready():
-            logger.debug("Replica set for units: %s, is not ready", self._unit_ips)
-            self.unit.status = BlockedStatus("failed to initialise replica set")
+    def _reconfigure(self, event: ops.charm.RelationEvent) -> None:
+        """Adds/removes RelationEvent triggering unit from the replica set.
+
+        Note: Removal funcationality will be implemented in the following PR.
+
+        Args:
+            event: The triggering relation event.
+        """
+        # check if reconfiguration is necessary
+        if self._need_replica_set_reconfiguration:
+            try:
+                # reconfigure replica set
+                self._mongo.reconfigure_replica_set()
+                # update the set of replica set hosts
+                self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
```

Comment:
The second comment is useful as it provides an interpretation to the code. By contrast, the first comment just repeats the code and is thus redundant.